mixer/pkg/runtime/dispatcher.go

```diff
@@ -281,8 +280,32 @@ func (m *dispatcher) Quota(ctx context.Context, requestBag attribute.Bag,
 // Preprocess runs the first phase of adapter processing before any other adapters are run.
 // Attribute producing adapters are run in this phase.
 func (m *dispatcher) Preprocess(ctx context.Context, requestBag attribute.Bag, responseBag *attribute.MutableBag) error {
-	// FIXME
-	return errors.New("not implemented")
+	_, err := m.dispatch(ctx, requestBag, adptTmpl.TEMPLATE_VARIETY_ATTRIBUTE_GENERATOR,
+		func(call *Action) []dispatchFn {
+			ra := make([]dispatchFn, 0, len(call.instanceConfig))
+			for _, inst := range call.instanceConfig {
+				ra = append(ra,
+					func(ctx context.Context) *result {
+						mBag, err := call.processor.ProcessGenAttrs(ctx, inst.Name,
+							inst.Params.(proto.Message),
+							requestBag, m.mapper,
+							call.handler)
+						if err == nil {
+							if err = responseBag.Merge(mBag); err != nil {
```

Comment:
Is Merge a thread safe method ?
This will be called from multiple threads at the same time.

Reply:
Option 1 --
Just add a local lock on line 283 in the "Preprocess" func and use that lock to externally protect responseBag.Merge mehod.

Options 2:

`mBagChan := make(chan mbag, 10)`    --> choose the size correctly.

`if err == nil {
     mBagChan <- mbag
}`

m.dispatch call in synchronous, so we know with certainty that all processing has been done by line 305.
Now you can range over the channel.

I think 1 is better.